pkg/agent/controller/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,879 @@
+			return err
+
+		}
+
+		pcState.maxNumCapturedPackets = pc.Spec.CaptureConfig.FirstN.Number
+		file, writer, err := getPacketFileAndWriter(pc.Name)
+		if err != nil {
+			return err
+		}
+
+		pcState.pcapngFile = file
+		pcState.pcapngWriter = writer
+		pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+		timeout := defaultTimeoutDuration
+		if pc.Spec.Timeout != nil {
+			timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+		}
+
+		err = c.performCapture(pcState, timeout, packetSource)
+		func() {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			c.captures[name] = "COMPLETED"
+			c.numCaptures -= 1
+			c.cond.Signal()
+		}()
+
+		return err
+	}()
+
+}
+
+func (c *Controller) performCapture(captureState *packetCaptureState, timeout time.Duration, source PacketSource) error {
+	options := CaptureOptions{
+		MaxCaptureLength: snapshotLen,
+		Promiscuous:      true,
+	}
+	packets, err := source.Capture(&options)
+	if err != nil {
+		klog.ErrorS(err, "failed to start capture")
+		return err
+	}
+
+	timer := time.NewTicker(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case packet := <-packets:
+			if captureState.numCapturedPackets == captureState.maxNumCapturedPackets {
+				return nil
+			}
+			captureState.numCapturedPackets++
+			ci := gopacket.CaptureInfo{
+				Timestamp:     time.Now(),
+				CaptureLength: len(packet.Data()),
+				Length:        len(packet.Data()),
+			}
+			err = captureState.pcapngWriter.WritePacket(ci, packet.Data())
+			if err != nil {
+				return fmt.Errorf("couldn't write packet: %w", err)
+			}
+			klog.V(5).InfoS("capture packet", "name", captureState.name, "count",
+				captureState.numCapturedPackets, "len", ci.Length)
+
+			reachTarget := captureState.numCapturedPackets == captureState.maxNumCapturedPackets
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
+				}
+				// if reach the target. flush the file and upload it.
+				if reachTarget {
+					path := os.Getenv("POD_NAME") + ":" + nameToPath(pc.Name)
+					if err := captureState.pcapngWriter.Flush(); err != nil {
+						return err
+					}
+					if pc.Spec.FileServer != nil {
+						err := c.uploadPackets(pc, captureState.pcapngFile)
+						klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+						// update upload result.
+						if updateErr := c.updatePacketCaptureStatus(pc.Name, captureState.numCapturedPackets, path, err); updateErr != nil {
+							return updateErr
+						}
+						if err != nil {
+							return err
+						}
+					} else {
+						// update capture result.
+						if updateErr := c.updatePacketCaptureStatus(pc.Name, captureState.numCapturedPackets, path, nil); updateErr != nil {
+							return updateErr
+						}
+					}
```

Comment:
One question about the `path`, it seems the path info will always be the one in the Pod, how do you show the uploaded status in the file server?

```suggestion
					if pc.Spec.FileServer != nil {
						err := c.uploadPackets(pc, captureState.pcapngFile)
						klog.V(4).InfoS("Uploaded captured packets", "name", pc.Name, "path", path)
					} 
					// update capture result.
					if updateErr := c.updatePacketCaptureStatus(pc.Name, captureState.numCapturedPackets, path, err); updateErr != nil {
							return updateErr
					}
					if err != nil {
							return err
					}
					
```

Reply:
we have a condition for that. `PacketCaptureFileUploaded`.  The working antrea-agent will perform the upload and update status.